job_list_params.go

```diff
@@ -0,0 +1,288 @@
+package river
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/riverqueue/river/internal/dbadapter"
+	"github.com/riverqueue/river/rivertype"
+)
+
+// JobListCursor is used to specify a starting point for a paginated
+// job list query.
+type JobListCursor struct {
+	id    int64
+	kind  string
+	queue string
+	time  time.Time
+}
+
+// JobListCursorFromJob creates a JobListCursor from a JobRow.
+func JobListCursorFromJob(job *rivertype.JobRow) *JobListCursor {
+	return &JobListCursor{
+		id:    job.ID,
+		kind:  job.Kind,
+		queue: job.Queue,
+		time:  jobListTimeValue(job),
+	}
+}
+
+// UnmarshalText implements encoding.TextUnmarshaler to decode the cursor from
+// a previously marshaled string.
+func (c *JobListCursor) UnmarshalText(text []byte) error {
+	dst := make([]byte, base64.StdEncoding.DecodedLen(len(text)))
+	n, err := base64.StdEncoding.Decode(dst, text)
+	if err != nil {
+		return err
+	}
+	dst = dst[:n]
+
+	wrapperValue := jobListPaginationCursorJSON{}
+	if err := json.Unmarshal(dst, &wrapperValue); err != nil {
+		return err
+	}
+	*c = JobListCursor{
+		id:    wrapperValue.ID,
+		kind:  wrapperValue.Kind,
+		queue: wrapperValue.Queue,
+		time:  wrapperValue.Time,
+	}
+	return nil
+}
+
+// MarshalText implements encoding.TextMarshaler to encode the cursor as an
+// opaque string.
+func (c JobListCursor) MarshalText() ([]byte, error) {
+	wrapperValue := jobListPaginationCursorJSON{
+		ID:    c.id,
+		Kind:  c.kind,
+		Queue: c.queue,
+		Time:  c.time,
+	}
+	data, err := json.Marshal(wrapperValue)
+	if err != nil {
+		return nil, err
+	}
+	dst := make([]byte, base64.URLEncoding.EncodedLen(len(data)))
+	base64.URLEncoding.Encode(dst, data)
+	return dst, nil
+}
+
+type jobListPaginationCursorJSON struct {
+	ID    int64     `json:"id"`
+	Kind  string    `json:"kind"`
+	Queue string    `json:"queue"`
+	Time  time.Time `json:"time"`
+}
+
+// SortOrder specifies the direction of a sort.
+type SortOrder int
+
+const (
+	// SortOrderAsc specifies that the sort should in ascending order.
+	SortOrderAsc SortOrder = iota
+	// SortOrderDesc specifies that the sort should in descending order.
+	SortOrderDesc
+)
+
+// JobListOrderByField specifies the field to sort by.
+type JobListOrderByField int
+
+const (
+	// JobListOrderByTime specifies that the sort should be by time. The specific
+	// time field used will vary by job state.
+	JobListOrderByTime JobListOrderByField = iota
```

Comment:
Are you sure about this one? Wouldn't it be better just to have an option for each timestamp field that you'd specify explicitly for what to sort on?

Currently, you have to read source code to understand what this will order by, and even if it were to be documented, it might be faster/easier to have users explicitly specify the timestamp they want.

Reply:
I'm reasonably sure that this is the behavior I want in the UI, though I could easily extract the logic that maps state to sort field into that external library. On the other hand, it'd be easy for us to add additional field-specific sort options in here as well. Part of the issue is that it doesn't necessarily make sense to sort by i.e. `finalized_at` when listing jobs that aren't yet finalized, so you'd get a bunch of jobs with `NULLS FIRST` at the beginning of the list. You could default to `scheduled_at` but that doesn't seem particularly relevant when listing jobs that have already completed.

Do you specifically think this behavior is bad and confusing to the extent that it shouldn't be included at all, or that we should have a different default, or better documentation, or just that we shouldn't ship this without additional options?

Reply:
> Do you specifically think this behavior is bad and confusing to the extent that it shouldn't be included at all, or that we should have a different default, or better documentation, or just that we shouldn't ship this without additional options?

Oh we should definitely have list ordering behavior included — I was thinking more the difference of making it explicit versus implicit. Implicit would be more like you have here where you tell it to sort by time and you get back certain behaviors depending on context. Explicit would be more like you specifically ask to sort on `ScheduledAt`, `FinalizedAt`, etc.

The documentation could be improved, but IMO a case could still be made that this can still be slightly suboptimal because you need to read the docs to know what it's going to do instead of it being pretty obvious.

I see what you mean that this is definitely a sensible thing from the UI's perspective, although I think there's a chance that this should behave more like a dumb ORM and require everything to be explicit from the programmatic API here.

All that said, I guess explicit order by fields could be added in the future and cohabitate peacefully with this one, so having it available isn't harmful. Not strongly against if you want to go this direction.